debug/debug.go

```diff
@@ -0,0 +1,380 @@
+	EnvVarKopiaDebugPprof = "KOPIA_DEBUG_PPROF"
+)
+
+// flags used to configure profiling in EnvVarKopiaDebugPprof.
+const (
+	// KopiaDebugFlagForceGc force garbage collection before dumping heap data.
+	KopiaDebugFlagForceGc = "forcegc"
+	// KopiaDebugFlagDebug value of the profiles `debug` parameter.
+	KopiaDebugFlagDebug = "debug"
+	// KopiaDebugFlagRate rate setting for the named profile (if available). always an integer.
+	KopiaDebugFlagRate = "rate"
+)
+
+const (
+	// ProfileNameBlock block profile key.
+	ProfileNameBlock ProfileName = "block"
+	// ProfileNameMutex mutex profile key.
+	ProfileNameMutex = "mutex"
+	// ProfileNameCPU cpu profile key.
+	ProfileNameCPU = "cpu"
+)
+
+// ProfileConfig configuration flags for a profile.
+type ProfileConfig struct {
+	flags []string
+	buf   *bytes.Buffer
+}
+
+// ProfileConfigs configuration flags for all requested profiles.
+type ProfileConfigs struct {
+	mu  sync.Mutex
+	pcm map[ProfileName]*ProfileConfig
+}
+
+//nolint:gochecknoglobals
+var pprofConfigs = &ProfileConfigs{}
+
+type pprofSetRate struct {
+	fn  func(int)
+	def int
+}
+
+//nolint:gochecknoglobals
+var pprofProfileRates = map[ProfileName]pprofSetRate{
+	ProfileNameBlock: {
+		fn:  func(x int) { runtime.SetBlockProfileRate(x) },
+		def: DefaultDebugProfileRate,
+	},
+	ProfileNameMutex: {
+		fn:  func(x int) { runtime.SetMutexProfileFraction(x) },
+		def: DefaultDebugProfileRate,
+	},
+}
+
+// GetValue get the value of the named flag, `s`.  false will be returned
+// if the flag value does not exist.
+func (p ProfileConfig) GetValue(s string) (string, bool) {
+	for _, f := range p.flags {
+		kvs := strings.SplitN(f, "=", pair)
+		if kvs[0] != s {
+			continue
+		}
+
+		if len(kvs) == 1 {
+			return "", true
+		}
+
+		return kvs[1], true
+	}
+
+	return "", false
+}
+
+func parseProfileConfigs(bufSizeB int, ppconfigs string) map[ProfileName]*ProfileConfig {
+	pbs := map[ProfileName]*ProfileConfig{}
+	allProfileOptions := strings.Split(ppconfigs, ":")
+
+	for _, profileOptionWithFlags := range allProfileOptions {
+		// of those, see if any have profile-specific settings
+		profileFlagNameValuePairs := strings.SplitN(profileOptionWithFlags, "=", pair)
+		flagValue := ""
+
+		if len(profileFlagNameValuePairs) > 1 {
+			flagValue = profileFlagNameValuePairs[1]
+		}
+
+		flagKey := ProfileName(strings.ToLower(profileFlagNameValuePairs[0]))
+		pbs[flagKey] = newProfileConfig(bufSizeB, flagValue)
+	}
+
+	return pbs
+}
+
+// newProfileConfig create a new profiling configuration.
+func newProfileConfig(bufSizeB int, ppconfig string) *ProfileConfig {
+	q := &ProfileConfig{
+		buf: bytes.NewBuffer(make([]byte, 0, bufSizeB)),
+	}
+
+	flgs := strings.Split(ppconfig, ",")
+	if len(flgs) > 0 && flgs[0] != "" { // len(flgs) > 1 && flgs[0] == "" should never happen
+		q.flags = flgs
+	}
+
+	return q
+}
+
+func setupProfileFractions(ctx context.Context, profileBuffers map[ProfileName]*ProfileConfig) {
```

Comment:
if you could make this a type `profileBuffers map[ProfileName]*ProfileConfig`, then `setupProfileFractions()` could be the called right after `parseProfileConfigs()`

Reply:
Similarly this seems stylistic ... I'd as that we postpone those changes until the PR is in and/or there's a concrete justification.